gtk/tests_backend/widgets/base.py

```diff
@@ -93,28 +110,66 @@ def shrink_on_resize(self):
 
     @property
     def color(self):
-        sc = self.native.get_style_context()
-        return toga_color(sc.get_property("color", sc.get_state()))
+        if GTK_VERSION < (4, 0, 0):  # pragma: no-cover-if-gtk4
+            sc = self.native.get_style_context()
+            return toga_color(sc.get_property("color", sc.get_state()))
+        else:  # pragma: no-cover-if-gtk3
+            style_provider = self.impl.style_providers.get(("color", id(self.native)))
+            style_value = (
+                style_provider.to_string().split(": ")[1].split(";")[0]
+                if style_provider
+                else None
+            )
+            return toga_color(style_value) if style_value else None
 
     @property
     def background_color(self):
-        sc = self.native.get_style_context()
-        return toga_color(sc.get_property("background-color", sc.get_state()))
+        if GTK_VERSION < (4, 0, 0):  # pragma: no-cover-if-gtk4
+            sc = self.native.get_style_context()
+            return toga_color(sc.get_property("background-color", sc.get_state()))
+        else:  # pragma: no-cover-if-gtk3
+            style_provider = self.impl.style_providers.get(
+                ("background_color", id(self.native))
+            )
+            style_value = (
+                style_provider.to_string().split(": ")[1].split(";")[0]
+                if style_provider
+                else None
+            )
+            return toga_color(style_value) if style_value else None
 
     @property
     def font(self):
-        sc = self.native.get_style_context()
-        return sc.get_property("font", sc.get_state())
+        if GTK_VERSION < (4, 0, 0):  # pragma: no-cover-if-gtk4
+            sc = self.native.get_style_context()
+            return sc.get_property("font", sc.get_state())
+        else:  # pragma: no-cover-if-gtk3
+            style_provider = self.impl.style_providers.get(("font", id(self.native)))
+            font_value = style_provider.to_string() if style_provider else None
+            return toga_font(font_value) if font_value else None
 
     @property
     def is_hidden(self):
         return not self.native.get_visible()
 
     @property
     def has_focus(self):
-        return self.native.has_focus()
+        if GTK_VERSION < (4, 0, 0):  # pragma: no-cover-if-gtk4
+            return self.native.has_focus()
+        else:  # pragma: no-cover-if-gtk3
+            root = self.native.get_root()
+            focus_widget = root.get_focus()
+            if focus_widget:
+                if focus_widget == self.native:
+                    return self.native.has_focus()
+                else:
+                    return focus_widget.is_ancestor(self.native)
```

Comment:
I presume this is because some GTK widgets are composed as multiple widgets and the actual focus is held  higher in the GTK widget tree than Toga might consider?